File: Backup.Web.Api.Server/Models/StockItem.cs
using System;
using Backup.Web.Api.Server.Services.Audit;

namespace Backup.Web.Api.Server.Models
{
	public class StockItem : Backup.Web.Api.Server.Services.Tenancy.IHasCompanyId, IHasAuditTrail
	{
		public int Id { get; set; }
		public string ProductKey { get; set; } = string.Empty; // Prefer ProductCode, else Product name
		public decimal QuantityOnHand { get; set; }
		/// <summary>P4 — quantité réservée par commandes confirmées (ATP = OnHand − Reserved).</summary>
		public decimal ReservedQuantity { get; set; }
		/// <summary>P4 — seuil de réappro auto (0 = désactivé).</summary>
		public decimal MinStock { get; set; }
		/// <summary>CMUP / CMP — coût moyen unitaire pondéré (société). Mis à jour à chaque entrée valorisée.</summary>
		public decimal AverageCost { get; set; }
		public DateTime LastUpdated { get; set; } = DateTime.UtcNow;
		/// <summary>
		/// ID du dernier BL qui a mis à jour ce produit.
		/// Permet une traçabilité rapide.
		/// </summary>
		public int? LastDeliveryId { get; set; }
		/// <summary>
		/// Fournisseur du produit (extrait du BL)
		/// </summary>
		public string? Supplier { get; set; }
		/// <summary>
		/// Description/libellé du produit
		/// </summary>
		public string? Description { get; set; }
		/// <summary>
		/// Unité du produit (ST, KG, PC, etc.)
		/// </summary>
		public string? Unit { get; set; }
		public string? CompanyId { get; set; }

		public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
		public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
		public string? CreatedBy { get; set; }
		public string? UpdatedBy { get; set; }
	}
}

File: backup.web.api.client/src/app/models/stock-item.ts
export interface StockItem {
  id: number;
  productKey: string;
  quantityOnHand: number;
  reservedQuantity?: number;
  minStock?: number;
  /** CMUP / CMP — coût moyen unitaire pondéré */
  averageCost?: number;
  lastUpdated: string; // ISO date string
  lastDeliveryId?: number | null; // ID du dernier BL qui a mis à jour ce produit
  supplier?: string | null; // Fournisseur
  description?: string | null; // Libellé du produit
  unit?: string | null; // Unité (ST, KG, PC, etc.)
}

export type StockForecastRisk = 'Critical' | 'Warning' | 'Watch' | 'Ok';
export type StockForecastTrend = 'Up' | 'Down' | 'Stable';

export interface StockForecastLine {
  stockItemId: number;
  productKey: string;
  description?: string | null;
  supplier?: string | null;
  quantityOnHand: number;
  reservedQuantity: number;
  available: number;
  minStock: number;
  qtyOutLookback: number;
  avgDailyOut: number;
  daysOfCover: number | null;
  dynamicMin: number;
  suggestedQty: number;
  risk: StockForecastRisk;
  trend: StockForecastTrend;
  stockoutAt?: string | null;
}

export interface StockForecastResult {
  lookbackDays: number;
  horizonDays: number;
  criticalCount: number;
  warningCount: number;
  watchCount: number;
  items: StockForecastLine[];
}

export interface StockUpdate {
  id: number;
  productKey: string;
  quantityDelta: number;
  quantityAfter: number;
  deliveryId: number;
  invoiceId?: number | null;
  updatedAt: string; // ISO date string
}

